src/wazuh.py

```diff
@@ -353,3 +354,57 @@ def _generate_cluster_snippet(
             <disabled>no</disabled>
         </cluster>
     """
+
+
+def change_api_password(username: str, old_password: str, new_password: str) -> None:
+    """Change Wazuh's API password for the default 'wazuh' user.
+
+    Args:
+        username: the username to change the user for.
+        old_password: the old API password for the user.
+        new_password: the new API password for the user.
+
+    Raises:
+        WazuhInstallationError: if an error occurs while processing the requests.
+    """
+    # The certificates might be self signed and there's no security hardening in
+    # passing them to the request since tampering with `localhost` would mean the
+    # container filesystem is compromised
+    try:
+        r = requests.get(  # nosec
+            "https://localhost:55000/security/user/authenticate",
+            auth=(username, old_password),
+            timeout=10,
+            verify=False,
+        )
+        # The old password has already been changed. Nothing to do.
+        if r.status_code == 401:
+            return
```

Comment:
Couldn't this be misleading ? How the caller will know if the password the passed is the one currently applied to the user ?
If the new password can't be applied, they should use the old password, but returning none doesn't allow to know that

Reply:
We don't return the old password. That's not the intent here. The password change should always happen and will occur only once, if the charm fails to change the password, reaching active status might result in a security issue.
If a 401 is returned, that means the password for that user is no longer the default one, it has already been changed. Hence, there's nothing else to do

Reply:
My concern is that this metod return nothing whether the password was successfully changed or not. So from the calling method perspective it's impossible to distinguish between a successful operation and a failed one.

If somewhere in the code I call a first time this method with "password1" and it succeeds
And the I call it a second time with "password2", and it fails because password will already be changed.
From the caller perspective on the second call I will think it succeeded, and will try to use "password2" which is wrong.

So the developer experience proposed by this method is misleading and could lead to issues in the future

Reply:
No, that scenario is not allowed, the password should not be changed for second time. That means "the I call it a second time with "password2", and it fails because password will already be changed" is considered successful because "password2" should have already been set.
If I'm understanding correctly, what you are proposing implies keeping track of the current password in the charm state, which increases significantly the attack surface. Storing passwords should be avoided whenever possible

Reply:
It's not because the scenario is not allowed, that someone can't call a second time this method with a different password when implementing a test or a new feature. Again the developer experience proposed here is misleading because the method will return nothing whether it succeeds or not.
In our take home test doing such return is considered as a disqualifying error, why applying something we consider bad ?

What I propose is throwing an exception is you get a 401 instead of silently returning. Because that's something that this method should surface, you can't change the password twice, if that ever happens, it's an error